research/delf/delf/python/training/global_features/train_utils.py

```diff
@@ -0,0 +1,368 @@
+
+        if (batch_num + 1) % update_every == 0 or (
+                batch_num + 1) == all_batch_num:
+          # Do one step for multiple batches. Accumulated gradients are
+          # used.
+          optimizer.apply_gradients(
+                  zip(accum_grads, model.trainable_variables))
+          accum_grads = [
+            tf.Variable(tf.zeros_like(tv.read_value()), trainable=False)
+            for tv in tvs]
+      except Exception as ex:
+        global_features_utils.debug_and_log(ex)
+        break
+
+    else:
+      # Validate one batch.
+      # We load full batch into memory.
+      input = []
+      target = []
+      try:
+        for _ in range(batch_size):
+          # Because the images are not necessarily of the same size, we can't
+          # set the batch size with .batch().
+          batch = loader.get_next()
+          input.append(batch[0:-1])
+          target.append(batch[-1])
+      except Exception as ex:
+        global_features_utils.debug_and_log(ex)
+        break
+
+      output = tf.zeros(shape=(0, model.meta['outputdim']), dtype=tf.float32)
+      for input_tuple in input:
+        for img in input_tuple:
+          # Compute the global descriptor vector.
+          model_out = model(tf.expand_dims(img, axis=0), training=False)
+          output = tf.concat([output, model_out], 0)
+
+      # No need to reduce memory consumption (no backward pass):
+      # Compute loss for the full batch.
+      tmp_target = tf.concat(target, axis=0)
+      queries = tf.boolean_mask(output, tmp_target == -1, axis=0)
+      positives = tf.boolean_mask(output, tmp_target == 1, axis=0)
+      negatives = tf.boolean_mask(output, tmp_target == 0, axis=0)
+      negatives = tf.reshape(negatives, [tf.shape(queries)[0], neg_num,
+                                         model.meta['outputdim']])
+      loss = criterion(queries, positives, negatives)
+
+      # Record loss.
+      losses.update(loss / batch_size, batch_size)
+
+    # Measure elapsed time.
+    batch_time.update(time.time() - end)
+    end = time.time()
+
+    # Record immediate loss and elapsed time.
+    if debug and ((batch_num + 1) % print_frequency == 0 or
+                  batch_num == 0 or (batch_num + 1) == all_batch_num):
+      global_features_utils.debug_and_log(
+              '>> {0}: [{1} epoch][{2}/{3} batch]\t Time val: {'
+              'batch_time.val:.3f} '
+              '(Batch Time avg: {batch_time.avg:.3f})\t Data {'
+              'data_time.val:.3f} ('
+              'Time avg: {data_time.avg:.3f})\t Immediate loss value: {'
+              'loss.val:.4f} '
+              '(Loss avg: {loss.avg:.4f})'.format(
+                      state, epoch, batch_num + 1, all_batch_num,
+                      batch_time=batch_time,
+                      data_time=data_time, loss=losses), debug=True, log=False)
+    batch_num += 1
+
+  return losses.avg
+
+
+def test(datasets, net, epoch, writer=None, model_directory=None,
```

Comment:
what about a more specific name? Like, test_retrieval? This would distinguish it from the validation step above which is about loss computation (which a reader could also intuitively think would be the case here)